ddtrace/contrib/langchain/patch.py

```diff
@@ -0,0 +1,713 @@
+                continue
+            self.metric(span, "dist", "tokens.%s" % token_type, num_tokens)
+        total_cost = span.get_metric("langchain.tokens.total_cost")
+        if total_cost:
+            self.metric(span, "incr", "tokens.total_cost", total_cost)
+
+
+def _extract_model_name(instance):
+    """Extract model name or ID from llm instance."""
+    for attr in ("model", "model_name", "model_id", "model_key", "repo_id"):
+        if hasattr(instance, attr):
+            return getattr(instance, attr, None)
+    return None
+
+
+def _extract_api_key(instance):
+    """
+    Extract and format LLM-provider API key from instance.
+    Note that langchain's LLM/ChatModel/Embeddings interfaces do not have a
+    standard attribute name for storing the provider-specific API key, so make a
+    best effort here by checking for attributes that end with `api_key/token`.
+    """
+    api_key_attrs = [a for a in dir(instance) if a.endswith(("api_token", "api_key"))]
+    if api_key_attrs and hasattr(instance, str(api_key_attrs[0])):
+        api_key = getattr(instance, api_key_attrs[0], None)
+        if api_key:
+            return "...%s" % api_key[-4:]
+    return ""
+
+
+def _tag_openai_token_usage(span, llm_output, propagated_cost=0, propagate=False):
+    """
+    Extract token usage from llm_output, tag on span.
+    Calculate the total cost for each LLM/chat_model, then propagate those values up the trace so that
+    the root span will store the total token_usage/cost of all of its descendants.
+    """
+    for token_type in ("prompt", "completion", "total"):
+        current_metric_value = span.get_metric("langchain.tokens.%s_tokens" % token_type) or 0
+        metric_value = llm_output["token_usage"].get("%s_tokens" % token_type, 0)
+        span.set_metric("langchain.tokens.%s_tokens" % token_type, current_metric_value + metric_value)
+    total_cost = span.get_metric("langchain.tokens.total_cost") or 0
+    if not propagate:
+        try:
+            completion_cost = get_openai_token_cost_for_model(
+                span.get_tag("langchain.request.model"),
+                span.get_metric("langchain.tokens.completion_tokens"),
+                is_completion=True,
+            )
+            prompt_cost = get_openai_token_cost_for_model(
+                span.get_tag("langchain.request.model"), span.get_metric("langchain.tokens.prompt_tokens")
+            )
+            total_cost = completion_cost + prompt_cost
+        except ValueError:
+            # If not in langchain's openai model catalog, assume 0 total cost.
+            pass
+    span.set_metric("langchain.tokens.total_cost", propagated_cost + total_cost)
+    if span._parent is not None:
+        _tag_openai_token_usage(span._parent, llm_output, propagated_cost=propagated_cost + total_cost, propagate=True)
+
+
+@with_traced_module
+def traced_llm_generate(langchain, pin, func, instance, args, kwargs):
+    llm_provider = instance._llm_type
```

Comment:
Should we first get the pin object and see if it's not None and enabled. We do this in most integrations. It might be best to keep the same pattern here. 

Ex from graphql:

```
pin = Pin.get_from(graphql)
if not pin or not pin.enabled():
    return func(*args, **kwargs)
```

Reply:
we use `@with_traced_module` which does this check internally (otherwise returns the wrapped function without running tracing code).